### DISCUSSION TITLE
Interpreter in Mojo?

### DISCUSSION THREAD

**Discussion by gryznar** [modular/modular#989] [Q&A]
I am wondering if Mojo runtime is based purely on JIT or interpreter will be also an option?

As far as I know, pure JIT is not the answer for every problem (e.g. when interpreting a code is much faster than compiling it and then running / interpreted code is easy in debugging / some tasks cannot be done via compilers etc.) and I am curious which way Mojo is following :)

Edit: This question may be answered somewhere, but I cannot find the answer.

Edit2: I am aware that it may be a newbie question, but I do not have experience with compiled languages to much and simply I don't know if it is possible to have the best from these 2 worlds (interpreted and compiled) without obvious drawbacks.

**Comment by PallHaraldsson** [comment]
I see in the docs:
>Mojo code can be ahead-of-time (AOT) or just-in-time (JIT) compiled.

This is the same, seemingly, as with Julia language, which isn't interpreted (by default, but also has as an interpreter as an opt in), uses JIT, or rather just-head-of-time (JAOT) compiled by default.

Both languages are interactive, as in have a REPL, and no bytecode (unlike Python that has bytecode and interactivity).

I do however see in the docs:
>Mojo REPL might hang (become unresponsive for more than 10 seconds) when interpreting an expression if your system has 4 GiB or less RAM.

This isn't traditional "interpreting" (an interpreter is a simple process and doesn't require 4 GB), likely the colloquial term from interactive /JIT.

As with Julia both languages use CPython for full Python (see PythonCall.jl how done for Julia). I.e. Mojo without CPython isn't a Python superset (and likely never will be, to me it very much seems like two languages in one), missing e.g. classes, but by calling CPython it adds support for Python code, and _then_ that code is (compiled to bytcode and it) interpreted, i.e. with no speedup.

You can take a look at discource.julialang.org. There are a few (long) discussions there on Mojo, e.g. this one on if Mojo is reinventing Julia (there are many similarities, they have similar speed, also some differences):

https://discourse.julialang.org/t/did-modular-just-reinvent-julia/98214

And I started this proposal a while back:
https://discourse.julialang.org/t/merge-python-and-julia-i-e-an-unofficial-integration-proposal-sort-of-like-the-mojo-language/101844

**Reply by gryznar** [reply]
Thanks for the answer!

Hah, if interpreter will be also an option in Mojo, it will be awesome! For development - interpreter, for production compiler.

**Reply by PallHaraldsson** [reply]
I don't officially speak for Mojo, only for what I think I knew and am digging up from the docs now. I would just nitpick the word "interpreted", but likely you have that, i.e. interactivity (what REPL implies) already. And you don't have to guess, you can check out already:

>You can now develop with Mojo locally. The Mojo SDK is currently available for Ubuntu Linux systems, and support for Windows and macOS is coming soon. You can still develop from a Windows or Mac computer using a container or remote Linux system.
>
>The Mojo SDK includes the Mojo standard library and the Mojo command-line interface (CLI), which allows you to run, compile, and package Mojo code. It also provides a REPL programming environment.

**Reply by gryznar** [reply]
I know :) I have SDK and I am not asking about "right now", but mainly about the future. I am also aware, that in REPL is JIT

**Reply by PallHaraldsson** [reply]
I'm confused, I'm talking about what I think is available right now. Since you have it (latest SDK version?), are you saying it's not interactive or has a REPL? Or that it only applies to Python code, not Mojo's addition? It's plausible and then I would want to know. [All languages can be compiled, just not to fast code, and all can be interpreted, so these terms do only apply to implementations (runtimes for interpreters), not languages. I'm not saying an interpreter will ever be planned (for Mojo additions that is), nor needed, just could be.]

**Reply by gryznar** [reply]
I would like to get some clarification from implementation details. Probably Mogball (compiler engineer at Modular) has mentioned somewhere, that Mojo is using interpreter for some tasks. I am just wondering in which areas it is used and if it is going to be an alternative runtime in the future :)

**Reply by Brian-M-J** [reply]
Chris Lattner has given an [answer](https://discord.com/channels/1087530497313357884/1152256775194673272/1154844887120883833) on how Mojo uses compilers and interpreters. Basically, Mojo uses an interpreter at compile time for metaprogramming and stuff, `mojo file.🔥` runs a JIT compiler, and `mojo build file.🔥` runs an AOT compiler.

JIT compilers are basically interpreters and AOT compilers fused together, to try to get the best of both worlds. For an explanation, I found [this video](https://youtu.be/sQTOIkOMDIw?si=-VJInQdJWTT2Rn-V) helpful.

**Reply by gryznar** [reply]
I have tested that and in some cases JIT is a problem. I've got script in which startup time is ~ 2s which causes, that relatively small Python code is faster (runs under 100ms)
